### ISSUE TITLE
Initialize more than once

### ISSUE THREAD

**Issue by Kretsch0** [gabime/spdlog#406]
Hi,

I got a problem by using log4cplus - the reason why I am watching to other loggers.
The problem is the initialization of the logger multiple times. For example through main and loaded dlls.
I looked to other frameworks like easylogging and they write:

`That simple! Please note that INITIALIZE_EASYLOGGINGPP should be used once and once-only otherwise you will end up getting compilation errors. This is definiting several extern variables. This means it can be defined only once per application. Best place to put this initialization statement is in file where int main(int, char**) function is defined, right after last include statement.`

So same result does not work. I read the easy example on GitHub about how to use the spdlog. And there is no init stuff to do. But does not mean there is no internal init that produce the same result?? Or problems at runtime?

So is it possible to log by using this header by using log functionality in many libraries and main for example?

**Comment by gabime** [comment]
Yes, there is a singlton (registry class), and I am not sure what would happen in the different scenarios involving with external dlls and libs..
I know of cases of spdlog being used in a external libraries but i never investigated the issue thoroughly. Would be great if if you could try and report your findings.

**Comment by noahdesu** [comment]
I would be interested in an update to this issue if there is any new information.

**Comment by yesint** [comment]
I'm running into problem when using spdlog in the library (Python extension module in my case).
The problem is that the singleton registry, which is created by C++ extension module, is only visible from the code of this extension module itself.
If I'm trying to access it from other extensions or from Python by means of wrappers I'm getting new empty instances of the registry instead.
I have no idea how to overcome this issue. How the registry gets initialized is completely obscure for me.

**Comment by paulsapps** [comment]
The way this usually works is something like:

```
[main exe]
[logger static lib, statically compiled]
       [other dll]
```

The singleton instance lives in the exe and the "other dll" can't see it. So usually the "fix" is that either [logger static lib, statically compiled] becomes a dll so its singleton is shared. Or that [other dll] also has the [logger static lib, statically compiled] with an exported function to set the singleton instance to point to the one in the main exe.

This is a bit wordy and wafflely so probably isn't easy to understand. 😂

**Comment by yesint** [comment]
@paulsapps could you please explain how to "set the singleton instance to point to the one in the main exe"? In my scenario the singleton exists in the shared library and have to be used in the other one. I can get its memory address, but what to do next?

**Comment by wang-bin** [comment]
Because spdlog is header only, each executable(dll,exe) MUST have it's own `s_instance`, i.e. it's not true singleton. I think the only solution is defining `instance()` in cpp and exporting the symbol for other libraries.

**Comment by paulsapps** [comment]
yeah what wang-bin said. That way when the library loads you can set its instance

**Comment by gabime** [comment]
The solution is to use the the compiled version of spdlog (1.4.0 or later)

**Comment by jakka30** [comment]
> The solution is to use the compiled version of spdlog (1.4.0 or later)
How can you fix this with static library? My goal is all my dlls and exe writing to one logger. How can I achieve this with static library? As you know, each dll and exe that links with the static library has its own copy of static library.
